### PR TITLE
Use OrdinalIgnoreCase for all Source and Package source mapping implementation

### LINKED ISSUE #11182: [Bug]: String comparison approach used for Source and Namespaces is not consistent
### NuGet Product Used

dotnet.exe, MSBuild.exe, NuGet.exe, Visual Studio Package Management UI, Visual Studio Package Manager Console

### Product Version

VS 2022 P3

### Worked before?

_No response_

### Impact

It bothers me. A fix would be nice

### Repro Steps & Context

Follow up regarding https://github.com/NuGet/NuGet.Client/pull/4212#discussion_r695943951

package source is using Ordinal comparison https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Items/SourceItem.cs#L80 whereas namespace Item is using OrdinalIgnoreCase comparison https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Items/PackageNamespacesSourceItem.cs#L211

there is another implementation for Package Source where keys are compared using `CurrentCultureIgnoreCase` approach. https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs#L148

### Verbose Logs

_No response_

### LINKED ISSUE #11241: [Bug]: VS crashes when package sources dropdown in PM UI has multiple package sources whose names are equal when compared using Culture Ignore Case
### NuGet Product Used

Visual Studio Package Management UI

### Product Version

VS 2022 and didn't test previous versions. 

### Worked before?

_No response_

### Impact

Other

### Repro Steps & Context

1. Create a `ClassLibrary` project in Visual Studio
2. Close the solution after it is created
3. Add a `nuget.config` file with below content under solution folder.
``` xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <config>
    <add key="globalPackagesFolder" value="globalPackagesFolder" />
  </config>
  <packageSources>
    <clear />
    <add key="encyclopaedia" value="https://api.nuget.org/v3/index.json" />
    <add key="encyclopædia" value="https://pkgs.dev.azure.com/azure-public/vside/_packaging/vssdk/nuget/v3/index.json" />
  </packageSources>  
</configuration>
```
> Note - The package source names are different when compared using ordinal ignore case and they are equal when compared using culture ignore case option. 

Reference - https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=net-5.0#examples

```
//       encyclopædia = encyclopaedia (CurrentCultureIgnoreCase): 
//       encyclopædia != encyclopaedia (OrdinalIgnoreCase): 
```
4. open the solution in VS 
5. Navigate to browse tab in PM UI
6. Change package source option to `All`

**Actual behavior:**
VS crashes

**Expected behavior:** 
VS opens packages in the browse tab successfully

[ClassLibrary38.zip](https://github.com/NuGet/Home/files/7181718/ClassLibrary38.zip)

vs-crash-report

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: https://github.com/nuget/home/issues/11182
Fixes: https://github.com/NuGet/Home/issues/11241

Regression? Last working version:

## Description
@kartheekp-ms  noticed we have inconsistent use of string comparison approach, after discussion we decided to use OrdinalIgnoreCase only for "Source and package source mapping", after my current changes I don't see any breaking or big difference after this change.

nuget.config used for my manual testing:

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="encyclopaedia" value="https://api.nuget.org/v3/index.json" />
    <add key="Encyclopaedia"  value="C:\Users\xxxx\1111" />
    <add key="encyclopædia" value="C:\Users\xxxx\2222" />
    <add key="Encyclopædia" value="https://pkgs.dev.azure.com/azure-public/vside/_packaging/vssdk/nuget/v3/index.json" />
  </packageSources>
</configuration>
```

I didn't do changes in following places due to breaking/unwanted change:

- See comment  https://github.com/NuGet/NuGet.Client/pull/4264#discussion_r709646405 https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs#L419

-  See comment https://github.com/NuGet/NuGet.Client/pull/4264#discussion_r709649975
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Items/SourceItem.cs#L80

In Apex tests I moved create nuget.config part before solution create action.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 1]
```diff
@@ -754,7 +754,7 @@ private ValueTask PopulateSourceRepoListAsync(IReadOnlyCollection<PackageSourceM
             {
                 SelectedSource = PackageSources
                     // if the old active source still exists. Keep it as the active source.
-                    .FirstOrDefault(i => StringComparer.CurrentCultureIgnoreCase.Equals(i.SourceName, selectedSourceName))
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:757]: This one decides current active source.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs [hunk 2]
```diff
@@ -416,7 +416,7 @@ private TryUpdateSourceResults TryUpdateSource()
                 return TryUpdateSourceResults.NotUpdated;
             }
 
-            if (selectedPackageSource.Name.Equals(name, StringComparison.CurrentCultureIgnoreCase) && selectedPackageSource.Source.Equals(source, StringComparison.OrdinalIgnoreCase))
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs:419]: I decided not to make this change in this file, revert the changes. It brings some unwanted side effects, if update existing source `encyclopaedia` with `Encyclopædia` on PMUI then save it creates duplicate entrees in nuget.config file.
Problem is when saving it check existing using packagesource key but due existing following code it assumes above 2 are different hence creates another entry instead of updating it. https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Items/SourceItem.cs#L80

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs [hunk 3]
```diff
@@ -425,7 +425,7 @@ private TryUpdateSourceResults TryUpdateSource()
 
             // check to see if name has already been added
             // also make sure it's not the same as the aggregate source ('All')
-            bool hasName = sourcesList.Any(ps => ps != selectedPackageSource && string.Equals(name, ps.Name, StringComparison.CurrentCultureIgnoreCase));
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs:428]: See https://github.com/NuGet/NuGet.Client/pull/4264/files#r708618554

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 4]
```diff
@@ -810,7 +810,7 @@ public string[] GetAvailableProjects()
                     safeAndDisplayName.Add(displayName);
                 }
                 // Sort with respect to the DisplayName
-                var sortedDisplayNames = safeAndDisplayName.OrderBy(i => i.Item2, StringComparer.CurrentCultureIgnoreCase).ToArray();
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: I'm reverting this one, because it's working project names, not related to source.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 5]
```diff
@@ -708,7 +708,7 @@ private void UpdateActiveSource(string activePackageSource)
             else
             {
                 var s = _packageSources.FirstOrDefault(
-                    p => StringComparer.CurrentCultureIgnoreCase.Equals(p, activePackageSource));
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:712]: This one let user to choose between `encyclopaedia` and `encyclopædia` sources. 
UI rendering not working correctly so created follow up issue https://github.com/NuGet/Home/issues/11235

File: src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs [hunk 6]
```diff
@@ -145,7 +145,7 @@ public bool Equals(PackageSource other)
                 return false;
             }
 
-            return Name.Equals(other.Name, StringComparison.CurrentCultureIgnoreCase) &&
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs]: After this change search is done on both `encyclopaedia and encyclopædia sources` as separate sources.

File: src/NuGet.Core/NuGet.PackageManagement/Utility/UriHelper.cs [hunk 7]
```diff
@@ -107,7 +107,7 @@ public static bool IsHttpSource(string source, PackageSourceProvider packageSour
                 }
 
                 var packageSource = packageSourceProvider.LoadPackageSources()
-                    .FirstOrDefault(p => p.Name.Equals(source, StringComparison.CurrentCultureIgnoreCase));
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.PackageManagement/Utility/UriHelper.cs:110]: This `IsHttpSource` method not used anywhere in our code, maybe we can remove it later.

File: src/NuGet.Core/NuGet.Configuration/Settings/ConfigurationDefaults.cs [hunk 8]
```diff
@@ -64,7 +64,7 @@ public IEnumerable<PackageSource> DefaultPackageSources
                         // In a SettingValue representing a package source, the Key represents the name of the package source and the Value its source
                         _defaultPackageSources.Add(new PackageSource(source.GetValueAsPath(),
                             source.Key,
-                            isEnabled: !disabledPackageSources.Any(p => p.Key.Equals(source.Key, StringComparison.CurrentCultureIgnoreCase)),
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Configuration/Settings/ConfigurationDefaults.cs:67]: It's about adding default sources from `NuGetDefaults.config`  which exist on `C:\Program Files (x86)\NuGet` path on Windows. https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior
Looks not popular feature, not sure how relevant now days.
Maybe we deprecate this one in the future?

File: src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs [hunk 9]
```diff
@@ -158,8 +158,8 @@ private static void AddDefaultPackageSources(List<PackageSource> loadedPackageSo
 
             foreach (var packageSource in defaultPackageSources)
             {
-                var sourceMatching = loadedPackageSources.Any(p => p.Source.Equals(packageSource.Source, StringComparison.CurrentCultureIgnoreCase));
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:161]: Related to https://github.com/NuGet/NuGet.Client/pull/4264/files#r708797119 on PMUI.
Looks not really relevant anymore.

File: src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs [hunk 10]
```diff
@@ -302,8 +302,8 @@ private PackageSource GetPackageSource(string key, Dictionary<string, IndexedPac
 
                 foreach (var packageSource in _configurationDefaultSources)
                 {
-                    var isSourceMatch = loadedPackageSources.Any(p => p.Source.Equals(packageSource.Source, StringComparison.CurrentCultureIgnoreCase));
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:305]: Related to https://github.com/NuGet/NuGet.Client/pull/4264/files#r708797119 for nuget.exe.
Triggered for `nuget sources Add -Name "MyServer" -Source \\myserver\packages` (https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-sources).
Looks not really relevant anymore.

File: src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ContextInfos/PackageSourceContextInfo.cs [hunk 11]
```diff
@@ -48,7 +48,7 @@ public bool Equals(PackageSourceContextInfo other)
                 return false;
             }
 
-            return Name.Equals(other.Name, StringComparison.CurrentCultureIgnoreCase) &&
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ContextInfos/PackageSourceContextInfo.cs]: After this change search both encyclopaedia and encyclopædia sources as separate sources.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs (hunk 2 above)
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs]: I decided not to make this change in this file, revert the changes. It brings some unwanted side effects, if update existing source `encyclopaedia` with `Encyclopædia` on PMUI then save it creates duplicate entrees in nuget.config file.
Problem is when saving it check existing using packagesource key but due existing key entry following code it assumes above 2 are different hence creates another entry instead of updating it.  Related to https://github.com/NuGet/NuGet.Client/pull/4264/commits/bdb3d6fe7a92a8fef615fc378afbfcf00ee29136#r709649975

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs (hunk 3 above)
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs]: I decided not to make this change in this file, revert the changes. It brings some unwanted side effects, if update existing source `encyclopaedia` with `Encyclopædia` on PMUI then save it creates duplicate entrees in nuget.config file.
Problem is when saving it check existing using packagesource key but due existing key entry following code it assumes above 2 are different hence creates another entry instead of updating it.  Related to https://github.com/NuGet/NuGet.Client/pull/4264/commits/bdb3d6fe7a92a8fef615fc378afbfcf00ee29136#r709649975

File: src/NuGet.Core/NuGet.Configuration/Settings/Items/SourceItem.cs [hunk 12]
```diff
@@ -77,8 +77,8 @@ public override bool Equals(object other)
                 return true;
             }
 
-            return string.Equals(Key, source.Key, StringComparison.Ordinal) &&
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Configuration/Settings/Items/SourceItem.cs]: I decided to revert this change.  For below input:
```
  <packageSources>
    <clear />
    <add key="encyclopaedia" value="C:\Users\eryondon\Desktop\NamespaceTestPackages\PublicRepository" />
    <add key="Encyclopaedia" value="https://api.nuget.org/v3/index.json" />
  </packageSources>
```
Before this change:
image

After this change:
image

Please note source url is quite different and looks breaking change experience for end user.

**Review by jeffkl — APPROVED** [review]

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs [hunk 13]
```diff
@@ -260,46 +254,43 @@ public async Task InstallPackageToProjectsFromUI_PackageNamespace_WithSingleFeed
...
 
             // Create nuget.config with Package namespace filtering rules before project is created.
             CommonUtility.CreateConfigurationFile(Path.Combine(solutionDirectory, "NuGet.config"), $@"<?xml version=""1.0"" encoding=""utf-8""?>
 <configuration>
     <packageSources>
     <!--To inherit the global NuGet package sources remove the <clear/> line below -->
     <clear />
-    <add key=""ExternalRepository"" value=""{externalRepositoryPath}"" />
-    <add key=""PrivateRepository"" value=""{privateRepositoryPath}"" />
+    <add key=""encyclopaedia"" value=""{externalRepositoryPath}"" />
```
Comment (erdembayar) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs:281]: Test "encyclopaedia" and "encyclopædia" sources used together doesn't crash VS PMUI setting.

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
Please review.

File: src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs (hunk 6 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs]: Were they treated the same before this change?

**Review by nkolev92 — COMMENTED** [review]
The fact that there's so many culture comparison gives me a pause about the impact this change could have.

File: src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs]: It looks it was not, so I'll revert changes here.
image

File: src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ContextInfos/PackageSourceContextInfo.cs (hunk 11 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ContextInfos/PackageSourceContextInfo.cs]: This one also looks irrelevant change, reverting because  it's already doing `Source` compare with `StringComparison.OrdinalIgnoreCase`. Here actually `Source` is more important than `Name`.

File: src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs]: Reverting because  it's already doing `Source` compare with `StringComparison.OrdinalIgnoreCase`. Here actually `Source` is more important than `Name`.

**Review by nkolev92 — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/LoadingStatusViewModel.cs [hunk 14]
```diff
@@ -22,7 +23,7 @@ public void UpdateModel(IItemLoaderState loaderState)
             PackageSearchStatus = Convert(loaderState.LoadingStatus);
             ItemsFound = loaderState.ItemsCount;
 
-            var convertedList = new System.Collections.SortedList();
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/LoadingStatusViewModel.cs:25]: Make VS stop crashing https://github.com/NuGet/Home/issues/11241
